### ISSUE TITLE
Never update the codebook parameters through gradients (only EMA is used)?

### ISSUE THREAD

**Issue by jzhang38** [lucidrains/vector-quantize-pytorch#40]
In many VQVAE papers (e.g., VQGAN, ViT-VQGAN), codebook loss (quantize.detach()-x) ** 2 is used to update the codebook embedding. This repo seems to abandon this loss completely and only updates the codebook through EMA. Is my understanding correct here?

**Comment by IISCAditayTripathi** [comment]
You either update using the codebook loss or through EMA. Both seems to work and newer implementations have moved on to EMA.

**Comment by lucidrains** [comment]
@IISCAditayTripathi yes that's correct, in fact, some newer works don't even rely on commitment loss for good results

**Comment by lucidrains** [comment]
@jzhang38 the commitment loss is for encouraging the features in your network to 'commit' to one of the codes, but the codes are updated through EMA
